076-moving-data-between-two-jbod-disks-using-cruise-control.md

```diff
@@ -0,0 +1,85 @@
+# Moving data between two JBOD disks using Cruise Control
+
+This proposal is about integrating the [`remove_disks`](https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/resources/yaml/endpoints/removeDisks.yaml) endpoint from Cruise Control into Strimzi cluster operator. 
+This endpoint will allow us to move the data between two JBOD disks. 
+
+## Current situation
+
+Currently, we get a multiple requests from community users to add the ability for moving all Kafka logs between two disks on the JBOD storage array. This feature can be useful in following scenarios:
+- The current disk is too small and the user wants to use a bigger one, or vice versa.
+- When we want to use a different Storage Class with different parameters or different storage types.
+- In case of disk removal to reduce the total storage.
+
+For now, we can do this using the Kafka CLI tool i.e. `kafka-reassign-partitions.sh` tool, but it takes a lot of manual steps which is time-consuming and not so user-friendly.
+
+## Motivation
+
+We should introduce the logic to Strimzi to leverage Cruise Control integration and make it possible to move the data between two JBOD disks.
+This feature will also allow us to remove the disks without the loss of data.
+
+## Proposal
+
+Cruise Control provides `remove_disks` HTTP REST endpoint to move replicas from a specified disk to other disks of the same broker. It always uses intra-broker re-balancing.
+This endpoint triggers a rebalancing operation by moving replicas in a size-based manner to the remaining disks, from the largest to the smallest, while checking the following constraint:
+```sh
+1 - (remainingUsageAfterRemoval / remainingCapacity) > errorMargin
```

Comment:
Out of curiosity ... Does Cruise Control take into account the sizes of individual partitions and the free space on individual disks? E.g. what happens in a situation when:
* I have broker with 3 disks, each 100Gi
* The disks 1 and 2 have each 1Gi of free space
* The 3rd disk hosts only one partition with 1,5Gi in size
* User requests to remove the 3rd disk
* => The overall free capacity is 2Gi, so the partition fits to that. But because each of the remaining two disks has only 1Gi of free space, it cannot be moved safely. So what happens? Does Cruise Control handle it and fail?

Reply:
Yes, looking at the code 
```
            double futureUsage = 0.0;
            double remainingCapacity = 0.0;
            for (Disk disk : broker.disks()) {
                futureUsage += disk.utilization();
                if (!logDirsToRemove.contains(disk.logDir())) {
                    remainingCapacity += disk.capacity();
                }
            }
            if (futureUsage / remainingCapacity > capacityThreshold) {
                LOG.error("Not enough remaining capacity to move replicas to for broker {}.", brokerId);
                throw new IllegalArgumentException(String.format("Not enough remaining capacity to move replicas to for broker %d.", brokerId));
            }
        }
```
If we exceed the capacity threshold then in that case, Cruise control will report not enough capacity

Reply:
But having multiple disks cannot be only dealt with through the threshold, or? Or what if the threshold is not big enough to detect the issue? In an extreme case, you can have something like this:
* Three 100Gi disks
* Three partitions, each 51Gi big, one on each disk
* What happens when you try to remove one of the disks? 
* You have 98Gi of free disk space on the remaining two disks. So the partitions from the third disk should seemingly fit in easily. But it will not.
 
Does Cruise Control give you a reasonable error and doesn't free the disk?

Reply:
I do see a check which makes sure that we are violating the disk capacity 
```

Reply:
I do see a check which makes sure that we are not violating the disk capacity while doing intra broker rebalancing
```
  /**
   * Check whether the movement of source replica to destination disk is acceptable for this goal.
   *
   * @param sourceReplica   Source replica.
   * @param destinationDisk Destination disk.
   * @return {@code true} if movement is acceptable for this goal, {@code false} otherwise.
   */
  private boolean isMovementAcceptableForCapacity(Replica sourceReplica, Disk destinationDisk) {
    double replicaUtilization = sourceReplica.load().expectedUtilizationFor(RESOURCE);
    return isUtilizationUnderLimitAfterAddingLoad(destinationDisk, replicaUtilization);
  }
```

Reply:
I n case the disk utilization is over limit we have this code handling the errors -> 
```
      for (Disk disk : broker.disks()) {
        if (disk.isAlive() && isUtilizationOverLimit(disk)) {
          // The utilization of the host for the resource is over the capacity limit.
          double requiredCapacity = disk.utilization() / _balancingConstraint.capacityThreshold(RESOURCE);
          ProvisionRecommendation recommendation = new ProvisionRecommendation.Builder(ProvisionStatus.UNDER_PROVISIONED)
              .numDisks(1).totalCapacity(requiredCapacity).build();
          throw new OptimizationFailureException(String.format("[%s] Utilization (%.2f) for disk %s on broker %d is above capacity limit.",
                                                               name(), disk.utilization(), disk, broker.id()), recommendation);
        }
        if (_shouldEmptyZeroCapacityDisks && disk.capacity() == 0 && !disk.replicas().isEmpty()) {
          throw new OptimizationFailureException(String.format("[%s] Could not move all replicas from disk %s on broker %d",
                                                               name(), disk.logDir(), broker.id()));
        }
      }
 ```

Reply:
Is that a yes or no for my question?

Reply:
From what I understood from the code, Yes

Reply:
Great, thanks.